pennylane/transforms/transpile.py

```diff
@@ -0,0 +1,156 @@
+        1: ──╰X──╭C──╭C──────│───├┤ Probs
+        2: ──╭C──│───╰X──╭C──│───├┤ Probs
+        3: ──╰X──╰X──────╰X──╰X──╰┤ Probs
+
+    Suppose we have a device which has connectivity constraints according to the graph:
+
+        0 --- 1
+        |     |
+        2 --- 3
+
+    We encode this in a coupling map as a list of the edges which are present in the graph, and then pass this, together
+    with the circuit, to the transpile function to get a circuit which can be executed for the specified coupling map
+
+    >>> dev = qml.device('default.qubit', wires=[0, 1, 2, 3])
+    >>> transpiled_circuit = qml.transforms.transpile(coupling_map=[(0, 1), (1, 3), (3, 2), (2, 0)])(circuit)
+    >>> transpiled_qnode = qml.QNode(circuit, dev)
+    >>> print(qml.draw(transpiled_qnode))
+
+     0: ──╭C─────────────────────╭C──╭┤ Probs
+     1: ──╰X──╭C─────────╭C──────│───├┤ Probs
+     2: ──╭C──│───╭SWAP──│───╭X──╰X──├┤ Probs
+     3: ──╰X──╰X──╰SWAP──╰X──╰C──────╰┤ Probs
+
+    A swap gate has been applied to wires 2 and 3, and the remaining gates have been adapted accordingly
+
+    """
+    # init connectivity graph
+    coupling_graph = nx.Graph(coupling_map) if not isinstance(coupling_map, nx.Graph) else coupling_map
+
+    # make sure every wire is present in coupling map
+    if any([wire not in coupling_graph.nodes for wire in tape.wires]):
+        raise ValueError(
+            f'not all wires present in coupling map! wires: {tape.wires}, coupling map: {coupling_graph.nodes}')
+
+    def _adjust_op_indices(_op, _map_wires):
+        """ helper function which adjusts wires in Operation according to the map _map_wires"""
+        _new_wires = Wires([_map_wires[w] for w in _op.wires])
+        _params = _op.parameters
+        if len(_params) == 0:
+            return type(_op)(wires=_new_wires)
+        return type(_op)(*_params, wires=_new_wires)
+
+    def _adjust_mmt_indices(_m, _map_wires):
+        """ helper function which adjusts wires in MeasurementProcess according to the map _map_wires"""
+        _new_wires = Wires([_map_wires[w] for w in _m.wires])
+
+        if _m.obs is None:
+            return type(_m)(return_type=_m.return_type, eigvals=_m.eigvals, wires=_new_wires)
+
+        # change wires of observable
+        _params = _m.obs.parameters
+        if len(_params) == 0:
+            _new_obs = type(_m.obs)(wires=_new_wires, id=_m.obs.id)
+        else:
+            _new_obs = type(_m.obs)(_params, wires=_new_wires, id=_m.obs.id)
+        return type(_m)(return_type=_m.return_type, obs=_new_obs)
+
+    gates = []
+
+    # we wrap all manipulations inside stop_recording() so that we don't queue anything due to unrolling of of templates
+    # or newly applied swap gates
+    with stop_recording():
+        # this unrolls everything in the current tape (in particular templates)
+        stop_at = lambda obj: (obj.name in all_ops) and (not getattr(obj, "only_visual", False))
+        expanded_tape = tape.expand(stop_at=stop_at)
```

Comment:
I'm curious as to why this was needed 🙂

Reply:
the purpose of this is so that we can access all ops inside parts of the circuit which are not unrolled (e.g. templates). Beacuse these parts can act on > 2 wires, the subsequent logic wouldn't apply to them and the overall circuit doesn't get transpiled correctly (i.e. these parts could not be executed on hardware).